internals/plan/plan.go

```diff
@@ -1021,82 +1022,111 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
+		fileName := finfo.Name()
+		if finfo.IsDir() || !strings.HasSuffix(fileName, ".yaml") {
 			continue
 		}
 		// TODO Consider enforcing permissions and ownership here to
 		//      avoid mistakes that could lead to hacks.
-		match := fnameExp.FindStringSubmatch(finfo.Name())
+		match := fnameExp.FindStringSubmatch(fileName)
 		if match == nil {
-			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", finfo.Name())
+			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", fileName)
 		}
 
-		data, err := ioutil.ReadFile(filepath.Join(dirname, finfo.Name()))
+		filePath := filepath.Join(dirname, fileName)
+		data, err := ioutil.ReadFile(filePath)
 		if err != nil {
 			// Errors from package os generally include the path.
 			return nil, fmt.Errorf("cannot read layer file: %v", err)
 		}
 		label := match[2]
-		order, err := strconv.Atoi(match[1])
+		localOrder, err := strconv.Atoi(match[1])
 		if err != nil {
 			panic(fmt.Sprintf("internal error: filename regexp is wrong: %v", err))
 		}
 
-		oldLabel, dupOrder := orders[order]
-		oldOrder, dupLabel := labels[label]
-		if dupOrder {
-			oldOrder = order
-		} else if dupLabel {
-			oldLabel = label
+		if prevFile, dupOrder := orders[localOrder]; dupOrder {
+			return nil, fmt.Errorf("invalid layer filename: %q not unique (have %q already with same order %d)",
+				fileName, prevFile, localOrder)
 		}
-		if dupOrder || dupLabel {
-			return nil, fmt.Errorf("invalid layer filename: %q not unique (have \"%03d-%s.yaml\" already)", finfo.Name(), oldOrder, oldLabel)
+		if prevFile, dupLabel := labels[label]; dupLabel {
+			return nil, fmt.Errorf("invalid layer filename: %q not unique (have %q already with same label %q)",
+				fileName, prevFile, label)
 		}
+		orders[localOrder] = fileName
+		labels[label] = fileName
+		lr.labels[label] = filePath
 
-		orders[order] = label
-		labels[label] = order
-
-		layer, err := ParseLayer(order, label, data)
+		// globalOrder of the layers must be maintained when using import paths.
+		globalOrder := lr.offset + localOrder
+		layer, err := ParseLayer(globalOrder, label, data)
 		if err != nil {
 			return nil, err
 		}
+		layer.Source = filePath
 		layers = append(layers, layer)
 	}
+	if len(layers) > 0 {
+		lr.offset = layers[len(layers)-1].Order + 1
+	}
 	return layers, nil
 }
 
-// ReadDir reads the configuration layers from the "layers" sub-directory in
+// Read reads the configuration layers from the "layers" sub-directory in
 // dir, and returns the resulting Plan. If the "layers" sub-directory doesn't
 // exist, it returns a valid Plan with no layers.
-func ReadDir(dir string) (*Plan, error) {
-	layersDir := filepath.Join(dir, "layers")
-	_, err := os.Stat(layersDir)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return &Plan{}, nil
+// If importDirs are passed, "layers" sub-directories of each of those
+// importDirs are read before the final dir.
```

Comment:
Hmm, this reads a bit oddly. It happens first, but it's at the end of the doc comment.

Could we just change the parameters to take a single `dirs []string` and have the caller append dir to the list it passes as the (single) argument?

Reply:
Fantastic idea, will make the change.